src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private int size;
+    private T[] elementData;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayFull();
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAddByIndex(index);
+        growIfArrayFull();
+        arrayCopyForAddByIndex(index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (elementData.length > list.size()) {
+            for (int index = 0; index < list.size(); index++) {
+                add(list.get(index));
+            }
+        } else {
+            grow();
+            addAll(list);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = elementData[index];
+        removeByIndex(elementData, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int index = 0; index < elementData.length; index++) {
+            if (equalsElemets(elementData[index], element)) {
+                removeByIndex(elementData, index);
+                return element;
+            }
+        }
+        throw new NoSuchElementException();
```

Comment:
If you throw an Exception, you should always add an informative message so that it would be clear from the stack trace what exactly went wrong.